### DISCUSSION TITLE
Will browser.getLogs() be available in WebdriverIO v9?

### DISCUSSION THREAD

**Discussion by thamp59** [webdriverio/webdriverio#13107] [Q&A] [answered]
Hi there,

I'm currently in the process of implementing a test that uses browser.getLogs('browser') in order to retrieve some of my console.logs in my chrome browser.

However, I see that the JSON Wire Protocol is being deprecated which allows for getLogs to be used.
![image](https://github.com/webdriverio/webdriverio/assets/92734962/4007022c-6906-4c74-86cc-12e0fb99d458)

Looking at the deprecation notes, I went and looked at devtools and while I was easily able to find devtools, I saw no implementation of getLogs. Furthermore I am unable to use this protocol because execution happens on sauce labs AND its being removed in V9 https://github.com/webdriverio/webdriverio/pull/12301
![image](https://github.com/webdriverio/webdriverio/assets/92734962/4c124058-baf9-4ab0-9eae-088005297f27)

My question is, will my test fallback to using ChromiumCommands.getLogs automatically when v9 hits or am I stuck not being able to use it?
![image](https://github.com/webdriverio/webdriverio/assets/92734962/bdb1b649-63a2-47bd-a4df-1c5efcb5a03d)

**Comment by christian-bromann** [comment, marked as answer]
> My question is, will my test fallback to using ChromiumCommands.getLogs automatically when v9 hits or am I stuck not being able to use it?

There will be two options for you:

- you will be able to use the [`@wdio/jsonwp-service`](https://www.npmjs.com/package/@wdio/jsonwp-service) which allows you to keep calling JSONWP commands also in upcoming versions of WebdriverIO
- you can use Bidi primitives like [`browser.sessionSubscribe(...)`](https://webdriver.io/docs/api/webdriverBidi#sessionsubscribe) to access browser logs, e.g.
```js
await browser.sessionSubscribe({ events: ['log.entryAdded'] })

/**
 * will print: {"type":"console","method":"log","realm":null,"args":[{"type":"string","value":"Hello Bidi"}],"level":"info","text":"Hello Bidi","timestamp":1657282076037}
 */
browser.on('log.entryAdded', (entryAdded) => console.log('received %s', entryAdded))

await browser.executeScript('console.log("Hello Bidi")', [])
```

I recommend to do the second suggestion and migrate to this API as I can't promise Chromium keeps this command around.

**Reply by thamp59** [reply]
Thanks! The second option led me down the path I actually wanted. I was trying to pull the browser logs to check for failing requests. Had a hard time getting the above implementation to grab the error logs related to the network requests (my assumption is that they are not actually logs, they just show up on the console, could be wrong though). I eventually ran into network.fetchError event.

`
await browser.sessionSubscribe({ events: ['network.fetchError'] })

browser.on('network.fetchError', (entry) => if (someCondition) console.log('received %s', entry))

await doSomeThingsInTheBrowser()

await browser.sessionUnsubscribe({ events: ['network.fetchError'] })
`

Sorry for not being able to accurately describe what I wanted. 🤣

**Reply by thamp59** [reply]
@christian-bromann Is it worth updating the docs to not suggest using "devtools" instead? https://webdriver.io/docs/api/jsonwp/#getlogs
If so, I can make an issue/contribution on this probably at some point.

**Reply by christian-bromann** [reply]
Any contributions would be awesome!
